### PR TITLE
fix unusual pads of conv

### PR DESCRIPTION
**Pull request by chanendy**
### 1. Background
     convert a Conv with pads[3,3,3,1] on onnx, than I receive an 'mismatch'

  image

### 2. Summary of corrections
     when there's Asymmetrical pads, consider it as user define params and lead it to 'get_padding_as_op'

### 5. Additional instructions
.This tools is really a great wok.
.English is not my native Language, sorry about that.
.I even do not how to send a model file for picture when PR.
.Hope this may somehow help this great work.

### CODE REVIEW

**Comment by PINTO0309** [comment]
Thanks for the suggestion. I need to see how it affects other models, so if possible please provide the model you are using.

Simply compress the file into a ZIP file and drag and drop it into the comment field for attachment.

example: [end2end.onnx.zip](https://github.com/PINTO0309/onnx2tf/files/12329464/end2end.onnx.zip)

**Comment by chanendy** [comment]
Here coms the model file

[padsConv.zip](https://github.com/PINTO0309/onnx2tf/files/12329491/padsConv.zip)

**Comment by PINTO0309** [comment]
Looks good. 

image

However, I have a feeling that the ONNX Pads are not the check you originally intended, as they have a slightly special way of holding values. 

https://github.com/onnx/onnx/blob/main/docs/Changelog.md#conv-11

image

The following is a partial quote from the ONNX specification text.

`pads` format should be as follow `[x1_begin, x2_begin...x1_end, x2_end,...]`, where xi_begin the number of pixels added at the beginning of axis `i` and xi_end, the number of pixels added at the end of axis `i`. This attribute cannot be used simultaneously with auto_pad attribute. If not present, the padding defaults to 0 along start and end of each spatial axis.

So, what I'm trying to say is,

sample.1:
ONNX: `pads: [3,3,3,1]` = TensorFlow: `pads: [[0,0], [3,3], [3,1], [0,0]]`

sample.2:
ONNX: `pads: [1,2,3,4]` = TensorFlow: `pads: [[0,0], [1,3], [2,4], [0,0]]`

File: onnx2tf/ops/Conv.py [hunk 1]
```diff
@@ -227,8 +227,18 @@ def make_node(
     # Check auto_pad nonexistent or NOTSET first
     pad_mode = 'VALID'
     padded = False
+
+    # for onnx pads that diff on the side of axes
+    # for example [3,1] or [3,3,3,1]
+    pads_axes_opposite_same = True
+    for i in range(0, spatial_size):
+        if pads[i * 2] != pads[i * 2 + 1]:
```
Comment (PINTO0309) [inline onnx2tf/ops/Conv.py:235]: Here.

Comment (PINTO0309) [reply onnx2tf/ops/Conv.py:235]: e.g.

```python
    for axis_begin, axis_end in zip(pads[0:spatial_size:1], pads[spatial_size::1]):
        if axis_begin != axis_end:
            pads_axes_opposite_same = False
            break
```

**Comment by chanendy** [comment]
You are right.

the model I send above is after optimized by this convert tool, maybe by the 'onnxsim' or some other graph optimization.

Here are the original code on Pytorch: (with `kernel.shape=[7,5]`)

`pad = (2, 0, 0, 0)`
`conv_pad = nn.ConstantPad2d(pad, 0.0)`
`chunked_inputs = conv_pad(x)`
`chunked_inputs = F.conv2d(`
`                  chunked_inputs,`
`                  kernel,`
`                  padding=[3, 1],`
 `)`

and the original  onnx model preview:

shortcut2


and the original model file:

[Ori_padsConv.zip](https://github.com/PINTO0309/onnx2tf/files/12329557/Ori_padsConv.zip)

**Comment by PINTO0309** [comment]
The behavior of the onnxsim looks correct. If you fix the logic and re-commit, I will merge it in after CI goes all green.

**Comment by chanendy** [comment]
Sure, your code is much more clear. I Just wonder how to make a re-commit.

**Comment by PINTO0309** [comment]
After modifying the code, you can execute a commit and it will automatically synchronize with GitHub.

**Comment by PINTO0309** [comment]
Okay, it will take about 2 hours for the regression test by CI, so please wait a bit.

**Comment by chanendy** [comment]
`After modifying the code, you can execute a commit + Push and it will automatically synchronize with GitHub.`

Is that mean another PR? 

I really new on these, you know, previously  we just merge code by hand.

By the way, Thanks for your patience (:

**Comment by PINTO0309** [comment]
> Is that mean another PR?

Your commit already overrides the first pull request.

image

https://github.com/PINTO0309/onnx2tf/pull/450/files

image

**Comment by chanendy** [comment]
All appreciation goes for you, this tools really help us a lot.
